// @microsoft/mixed-reality-extension-sdk/built/types/rigidBodyConstraints.js
"use strict";
/*!
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Flags to constrain rigid body motion.
 */
var RigidBodyConstraints;
(function (RigidBodyConstraints) {
    RigidBodyConstraints["None"] = "none";
    RigidBodyConstraints["FreezePositionX"] = "freeze-position-x";
    RigidBodyConstraints["FreezePositionY"] = "freeze-position-y";
    RigidBodyConstraints["FreezePositionZ"] = "freeze-position-z";
    RigidBodyConstraints["FreezePosition"] = "freeze-position";
    RigidBodyConstraints["FreezeRotationX"] = "freeze-rotation-x";
    RigidBodyConstraints["FreezeRotationY"] = "freeze-rotation-y";
    RigidBodyConstraints["FreezeRotationZ"] = "freeze-rotation-z";
    RigidBodyConstraints["FreezeRotation"] = "freeze-rotation";
    RigidBodyConstraints["FreezeAll"] = "freeze-all";
})(RigidBodyConstraints = exports.RigidBodyConstraints || (exports.RigidBodyConstraints = {}));
